torch/ao/quantization/experimental/observer.py

```diff
@@ -4,27 +4,134 @@
 """
 
 import torch
+import itertools
 from torch.ao.quantization.observer import ObserverBase
 
-class APoTObserver(ObserverBase):
-    alpha = 0
-    gamma = 0
-    level_indices = torch.Tensor()
+class NonUniformQuantizationObserverBase(ObserverBase):
+    min_val: torch.tensor([])
+    max_val: torch.tensor([])
+    level_indices: torch.tensor([])
+    b: int
+    k: int
+    n: int
+    alpha: float
+    gamma: float
 
     def __init__(
         self,
-        min_val: torch.Tensor,
-        max_val: torch.Tensor,
-        level_indices: torch.Tensor,
-            b: int,
-            k: int) -> None:
-        super().__init__
+        min_val=None,
+        max_val=None,
+        b=0,
+        k=0,
+            dtype=torch.quint8) -> None:
+        super().__init__(dtype)
+        self.min_val = min_val
+        self.max_val = max_val
+        self.level_indices = torch.tensor([])
+        self.b = b
+        self.k = k
+        self.n = 0
+        self.alpha = 0.0
+        self.gamma = 0.0
 
-    def calculate_qparams(self):
-        return self._calculate_qparams()
+    r""" Calculates nonuniform quantization parameters given min and max value tensors.
+    Args:
+        min_val: Minimum values per channel
+        max_val: Maximum values per channel
+    Returns:
+        gamma: gamma quantization parameter, defined to ensure that alpha is the maximum of the range
+        quantization_levels: non-uniform quantization levels, calculated according to APoT paper:
+        (https://arxiv.org/pdf/1909.13144.pdf)
+        level_indices: int representation of quantization_levels
+    """
+    def _calculate_qparams(
+        self,
+        min_val: torch.tensor,
+        max_val: torch.tensor,
+            signed: bool) -> Tuple[float, torch.tensor, torch.tensor]:
+        # compute alpha
+        self.alpha = max_val
+
+        # compute n and store as member variable
+        if self.k:
+            self.n = self.b // self.k
```

Comment:
your implementation should guarantee n is an integer by ensuring b is divisible by k. APoT doesn't make sense if b % k != 0